src/main/java/marquez/common/models/DatasourceUrn.java

```diff
@@ -1,25 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package marquez.common.models;
 
-import java.util.regex.Pattern;
-import lombok.EqualsAndHashCode;
+import static marquez.common.Preconditions.checkNotBlank;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
 import lombok.NonNull;
 
-@EqualsAndHashCode(callSuper = false)
 public final class DatasourceUrn extends Urn {
-  private static final int NUM_COMPONENTS = 2;
-  private static final String URN_TYPE = "datasource";
-  private static final Pattern REGEX = buildPattern(URN_TYPE, NUM_COMPONENTS);
+  private static final String NAMESPACE = "datasource";
+  private static final int NUM_OF_PARTS = 2;
+  private static final UrnPattern PATTERN = UrnPattern.from(NAMESPACE, NUM_OF_PARTS);
 
-  public DatasourceUrn(@NonNull String value) {
-    super(value, REGEX);
+  private DatasourceUrn(@NonNull final String value) {
+    super(checkNotBlank(value));
   }
 
-  public static DatasourceUrn from(@NonNull String type, @NonNull String name) {
-    final String value = fromComponents(URN_TYPE, type, name);
+  public static DatasourceUrn from(@NonNull DatasourceType type, @NonNull DatasourceName name) {
+    final String value = valueFrom(NAMESPACE, type.toString(), name.getValue());
     return fromString(value);
   }
 
+  @JsonCreator
   public static DatasourceUrn fromString(String value) {
     return new DatasourceUrn(value);
   }
+
+  @Override
+  public String namespace() {
+    return NAMESPACE;
```

Comment:
Can this be brought into the parent class, since it's implemented the same way here and also in `DatasetUrn`, both of which extend `Urn`?

Reply:
The abstract methods `namespace()` and `pattern()` forces anyone who extends the **class** `Urn` to implement those methods. Giving us a clear contract of how to interface with and define a URN.

So, yeah, there is some redundancy, but the current approach favors a more formal definition of a URN.